engine/src/main/java/org/hibernate/search/query/engine/impl/HSQueryImpl.java

```diff
@@ -670,7 +680,51 @@ else if ( this.sort != null && projection != null ) {
 		);
 	}
 
+	private void validate(Sort sort, Collection<Class<?>> classes) {
+		SortField[] sortFields = sort.getSort();
+		for ( SortField sortField : sortFields ) {
+			String field = sortField.getField();
+			boolean indexedField = false;
+			for ( Class<?> clazz : classes ) {
+				IndexedTypeDescriptor indexedTypeDescriptor = extendedIntegrator.getIndexedTypeDescriptor( clazz );
+				FieldDescriptor fieldDescriptor = indexedTypeDescriptor.getIndexedField( field );
+				if ( fieldDescriptor != null ) {
+					indexedField = true;
+					if ( sortField instanceof DistanceSortField ) {
+						validateDistanceSortField( field, clazz );
+					}
+					break;
+				}
+			}
+			if ( !indexedField ) {
+				throw log.cannotSortBecauseFieldIsNotIndexed( field );
+			}
+		}
+	}
 
+	private void validateDistanceSortField(String field, Class<?> clazz) {
+		EntityIndexBinding indexBinding = extendedIntegrator.getIndexBinding( clazz );
+		PropertyMetadata propertyMetadata = indexBinding.getDocumentBuilder().getMetadata().getPropertyMetadataForProperty( field );
+		if ( propertyMetadata != null ) {
+			// The field is a property of the entity
+			XClass elementClass = propertyMetadata.getPropertyAccessor().getElementClass();
```

Comment:
You shouldn't use the _XClass_ helpers here, as they will ignore other sources of metadata - for example the programmatic configuration.
Also, we shouldn't do reflection at runtime but somehow have information to validate this using our model.